stablehlo/dialect/Serialization.cpp

```diff
@@ -0,0 +1,89 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "stablehlo/dialect/Serialization.h"
+
+#include <sstream>
+
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Bytecode/BytecodeWriter.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "stablehlo/dialect/Version.h"
+#include "stablehlo/dialect/VhloOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+LogicalResult serializePortableArtifact(ModuleOp module,
+                                        std::string const& targetVersion,
+                                        raw_ostream& os) {
+  MLIRContext* context = module.getContext();
+
+  // Convert StableHLO --> VHLO. Will fail if entire program is not StableHLO.
+  {
+    mlir::PassManager pm(context);
+    pm.addPass(mlir::stablehlo::createStablehloLegalizeToVhloPass());
+    if (!mlir::succeeded(pm.run(module))) {
+      return failure();
+    }
+  }
+
+  // Convert VHLO --> VHLO(version x.y.z).
+  // Doing separately for now since we need to improve error messaging around
+  // target version failures.
+  {
+    mlir::PassManager pm(context);
+    pm.addPass(mlir::stablehlo::createVhloToVersionPass({targetVersion}));
+    if (!mlir::succeeded(pm.run(module))) {
+      return failure();
+    }
+  }
+
+  // Writes the header VHLO_<serializerVersion>_<targetVersion>
```

Comment:
What is the use case for storing this information, especially since it's not part of the deserialization API?

Reply:
It will help with resolving bugs if we know the serializer version. Target version is helpful as a repro aid.

Some email feedback I received recently:

> One other thought I had is how easy will be it to debug compatibility problems. Wouldn't it be useful for the serialized format to contain the date and code version of the serializer? Alternatively, we can have a table with the dates at which different VHLO versions were introduced, to make it easy to see if something was serialized too far in the past.

While I don't like storing the date (two artifacts serialized the same way by the same version should probably create a deterministic file), I think we should store some plaintext readable information for debugging.

Reply:
Removed header for now